Custom logger support

iatsiuk opened the pull request:
Hi,

This PR is from a discussion in [this](https://github.com/mixpanel/mixpanel-node/issues/221) issue.

It adds support to pass a custom logger to the mixpanel instance. The custom logger should implement the following interface:
```ts
interface CustomLogger {
  trace(message?: any, ...optionalParams: any[]): void;
  debug(message?: any, ...optionalParams: any[]): void;
  info(message?: any, ...optionalParams: any[]): void;
  warn(message?: any, ...optionalParams: any[]): void;
  error(message?: any, ...optionalParams: any[]): void;
}
```

Here is an example output for the console logger, which will be used by default if no logger is passed:
```
Sending the following data to Mixpanel (/engage) {
  data: {
    '$token': 'TOKEN',
    '$distinct_id': 'ID',
    '$set': {
      foo: 'bar'
    }
  }
}
```

Here is an example output for the [pinojs](https://github.com/pinojs/pino) logger:
```
[1692475025041] DEBUG (API/10184 on Nicolle.local): Sending the following data to Mixpanel (/engage) {"data":{"$token":"TOKEN","$distinct_id":"ID","foo":"bar"}}}
```

tdumitrescu commented on test/logger.js:
```diff
@@ -0,0 +1,100 @@
+        next();
+    },
+
+    "is default logger is console object": function(test) {
```
```suggestion
    "defaults to console logger": function(test) {
```

tdumitrescu commented on test/logger.js:
```diff
@@ -0,0 +1,100 @@
+        test.done();
+    },
+
+    "is throws an error on incorrect logger object": function(test) {
```
```suggestion
    "throws an error on incorrect logger object": function(test) {
```

tdumitrescu commented on test/logger.js:
```diff
@@ -0,0 +1,100 @@
+        test.done();
+    },
+
+    "is write log for track() method": function(test) {
```
```suggestion
    "writes log for track() method": function(test) {
```

tdumitrescu commented on test/logger.js:
```diff
@@ -0,0 +1,100 @@
+        test.done();
+    },
+
+    "is write log for increment() method": function(test) {
```
```suggestion
    "writes log for increment() method": function(test) {
```

tdumitrescu reviewed (commented):
Thanks so much, this is one of the cleanest PRs we've had come through on this repo. Just a couple requests around the tests:
- the suggested docstring changes are for consistency with the style of the rest of the tests: "[subject] [action]" -> "[logger] throws an error on incorrect logger object"
- the test cases currently only cover the default config (no custom logger). could you add a few cases where you pass in a custom logger (a stub/mock where you can assert that it's getting called, and NOT the default console logger)?

iatsiuk commented:
Thanks for your kind words. I will fix all the problems later in the week.

iatsiuk commented:
@tdumitrescu I have made the necessary corrections and addressed all the issues you raised. Please let me know if there is anything else I can do to improve the code or if you have any further suggestions.

tdumitrescu commented:
Thanks @iatsiuk. This has now been released in [v0.18.0](https://github.com/mixpanel/mixpanel-node/releases/tag/v0.18.0).
